### PR TITLE
Gracefully handle small datasets with too few contigs

### PR DESCRIPTION
**Pull request by paulzierep**
## Summary

Fix multiple failure modes when running COMEBin on small test datasets with too few contigs:

1. **DataLoader `drop_last=True` → `drop_last=False`**: When dataset is smaller than `batch_size`, `drop_last=True` drops every element, yielding an empty train loader and untrained model. Now uses `min(batch_size, len(train_dataset))` and `drop_last=False`.

2. **Replace `sys.exit()` with graceful fallback in `gen_seed()`**: When FragGeneScan or hmmsearch produce empty output on small contigs, the pipeline crashed via `sys.exit()`. Now returns `candK=0` so the pipeline continues gracefully.

3. **Guard HNSW index against <2 points**: `features[0]` raised `IndexError` with 0 elements. Now returns `None` early.

4. **Guard kNN query against k > available elements**: `knn_query(embeddings, k=101)` with only 1 element raised HNSWlib RuntimeError. Now caps `k` to `min(requested, len(embeddings))`.

5. **Guard seed k-means, Leiden clustering against too few nodes**: Empty marker seeds, `n_clusters < 2`, or `edgelist` all crashed. Added early-return guards.

6. **Guard training against 0 contigs**: Empty `train_dataset` now returns early instead of crashing on undefined `logits/labels`.

7. **Guard `gen_seed_idx` against missing/empty seed files**: Handles `FileNotFoundError` and empty seed lists gracefully.

## Testing

These changes prevent crashes on small datasets. No existing test infrastructure was available in the repo.

### CODE REVIEW

**Comment by paulzierep** [comment]
## Integration test command

This command runs COMEBin end-to-end on a small test dataset (50 contigs, ~100Kbp) to verify the small-dataset fixes work. It creates synthetic contigs and a minimal BAM file, then runs the full pipeline.

### Prerequisites

- Conda environment from `comebin_env.yaml` or a Python 3.9+ env with: torch, numpy, scipy, scikit-learn, biopython, samtools, hmmer, prodigal, pyyaml, tensorboard

### Test

```bash
# Create a small test dataset
python3 -c "
import os, random, subprocess
random.seed(42)
os.makedirs('test_data/bams', exist_ok=True)

# Generate 50 random contigs (1500-3000bp each)
with open('test_data/contigs.fasta', 'w') as f:
    for i in range(50):
        seq = ''.join(random.choice('ACGT') for _ in range(random.randint(1500, 3000)))
        f.write(f'>contig_{i}\\n{seq}\\n')

# Build a minimal BAM via samtools
subprocess.run(['samtools','faidx','test_data/contigs.fasta'], check=True)
ref_len = sum(int(l.split()[1]) for l in open('test_data/contigs.fasta.fai'))
with open('test_data/reads.sam', 'w') as f:
    f.write('@HD\\tVN:1.6\\tSO:unsorted\\n')
    f.write(f'@SQ\\tSN:genome\\tLN:{ref_len}\\n')
    for i, line in enumerate(open('test_data/contigs.fasta')):
        if line.startswith('>'):
            name = line[1:].strip()
            f.write(f'read_{i}\\t0\\t{name}\\t1\\t0\\t*\\t*\\t0\\t0\\t*\\t*\\n')
subprocess.run(['samtools','view','-bS','test_data/reads.sam','-o','test_data/bams/mapping.bam'], check=True)
subprocess.run(['samtools','sort','test_data/bams/mapping.bam','-o','test_data/bams/mapping.bam'], check=True)
subprocess.run(['samtools','index','test_data/bams/mapping.bam'], check=True)
print('Test data created in test_data/')
"

# Run the full pipeline
bash run_comebin.sh -a test_data/contigs.fasta \
                     -o test_data/output \
                     -p test_data/bams \
                     -t 4 \
                     -b 10 \
                     -n 6

# Check outputs
ls -la test_data/output/comebin_res/
ls -la test_data/output/comebin_res/cluster_res/
ls -la test_data/output/comebin_res/comebin_res_bins/ 2>/dev/null || echo "All bins filtered out (expected for tiny test data)"
```

**Note**: The `-b 10` flag sets batch size to 10 (smaller than the 50 contigs, so the `batch_size=min(...)` guard isn't triggered). Remove `-b 10` to test the default batch_size=1024 path where the DataLoader fix applies.

Expected outcome: All pipeline steps complete without crashing. Clustering produces .tsv files in `cluster_res/`. The `get_result` step picks the best Leiden result. `filter_small_bins` may filter all bins (min size 200Kbp) for such tiny data, which is correct behavior.

**Comment by paulzierep** [comment]
## Integration test: step-by-step

### 1. Create conda environment
```bash
conda env create -f comebin_env.yaml -n comebin_test
conda activate comebin_test
```

Depending on your CUDA version you may want to adjust the `pytorch` and `cudatoolkit` versions in the yaml. For CPU-only testing, replace `pytorch=1.10.2=py3.7_cuda11.1_cudnn8.0.5_0` and `cudatoolkit=11.1.1` with `pytorch=1.10.2` and remove `cudatoolkit`.

### 2. Clone the repo (PR branch)
```bash
git clone https://github.com/paulzierep/COMEBin.git
cd COMEBin
git checkout fix-small-test-data
```

### 3. Prepare input data
You need:
- **Contigs** in FASTA (at least a handful, but <1024 is fine for this test)
- **BAM files** (sorted, indexed) mapping reads back to the contigs

For a quick smoke test using a tiny assembly (e.g. 63 contigs from a real gut metagenome, available on request):
```bash
mkdir -p ~/comebin_test
cp /path/to/contigs.fasta ~/comebin_test/
cp /path/to/*.bam ~/comebin_test/bams/
```

### 4. Run the full pipeline
```bash
cd COMEBin
bash bin/run_comebin.sh \
  -a ~/comebin_test/contigs.fasta \
  -o ~/comebin_test/output \
  -p ~/comebin_test/bams/ \
  -t 4
```

### 5. Verify results
Pipeline succeeds if the final log line reads something like:
```
Final result: .../comebin_res/cluster_res/Leiden_bandwidth_0.05_res_....
```

Output bins are in:
```bash
ls -la ~/comebin_test/output/comebin_res/comebin_res_bins/
```

### CheckM fallback behavior
When CheckM marker gene tables are not available (e.g. because the dataset is too small for CheckM to identify markers, or CheckM is not installed properly):
1. The pipeline logs: `"CheckM marker gene tables not available. Using first available clustering result directly (skipping size filter)."`
2. It picks the first non-k-means clustering `.tsv` result from `cluster_res/` (sorted alphabetically).
3. It generates bins directly from that clustering without the 200kbp minimum-size filter, so even small datasets get output bins.
4. The unfiltered result is also written to `comebin_res.tsv`.

To force this fallback (e.g. to test it explicitly), run the `get_result` step with placeholder paths:
```bash
python main.py get_result \
  --contig_file ~/comebin_test/contigs.fasta \
  --output_path ~/comebin_test/output/comebin_res \
  --seed_file ~/comebin_test/contigs.fasta.bacar_marker.2quarter_lencutoff_1001.seed \
  --bac_mg_table /nonexistent \
  --ar_mg_table /nonexistent \
  --num_threads 4
```

### Summary of changes in this PR
| Commit | Fix |
|--------|-----|
| `84b56b3` | Gracefully handle datasets with <1024 contigs (early return, fallback clustering) |
| `ab4cbaf` | Unit tests for small-data edge cases |
| `65aea2e` | Fix `stable_cumsum` import for newer sklearn |
| `5bbd1fc` | sklearn compat: `n_jobs`, `algorithm='full'`, `np.int` |
| `0020319` | Compute `batch_size` from `features.shape[0] // n_views` in all 3 `info_nce_loss` variants |
| `92c5ac4` | Guard `cluster()` against missing seed file |
| `e566210` | Fix `run_comebin.sh` repo-root resolution (works with both `bash bin/run_comebin.sh` and `export PATH`) |
| `a69076c` | Improve CheckM fallback: generate bins directly without 200kbp filter |
